### ISSUE TITLE
creating config variables using save_config_variables() method is adding additional "\" character to password variables in config_variables.yml file

### ISSUE THREAD

**Issue by varundunga** [fivetran/great_expectations#2196]
**Describe the bug**
Im trying to create great expectations project from code instead of cli.  When creating config variables using save_config_variables() method, code is adding additional "\\" character to password variables in config_variables.yml file if password is environment variable.

**To Reproduce**
1. Create a great expectations data context object using ge.DataContext.create()
2. Create kwargs with all details about a connection where password is an environmental variable
2. pass connection name and connection kwargs to save_config_variables() method of data context object

**Expected behavior**
An entry should be created in config_variables.yml file inside uncommitted folder with all details present in kwargs  variable.
password value should not contain additional "\\" character. This works fine in version 0.12.4
Also, This works fine if password is not an environmental variable

**Environment (please complete the following information):**
 - OS:  Windows 10
 - GE Version: 0.12.8 and 0.13.3
**Additional context**
Code i used:
```
import great_expectations

kwargs = {
            "drivername": "snowflake",
            "host": "hostname",
            "warehouse": "compute_wh",
            "schema": "public",
            "database": "employee",
            "username": "username",
            "password": "${SF_PWD}"
            }
data_ctxt = ge.DataContext.create('./', False)
data_ctxt.save_config_variable('SF', kwargs)

```

**Comment by eugmandel** [comment]
The core team will take on reproducing and fixing this. Thank you for the report, @varundunga !

**Comment by anthonyburdi** [comment]
Hi @varundunga! Thank you for submitting this issue. The additional `\` character you are seeing is intended to be in the config file to escape the `$` character. We added this escaping in #2015 to make sure that we don't attempt to substitute values containing `$` characters in our configuration files. When reading these configurations via `context.get_config_with_variables_substituted()` the escaping is properly removed.

I wanted to make sure I didn't miss anything so I added an additional test in #2243 to explicitly test the workflow you provided, and also we noticed you are using Windows so I tested there as well.

FYI - I noticed you were using a `${SF_KEY}` in your example `config_variables`. Please note that environment variable substitution is not available in the `config_variables.yml` and only in the main data context configuration (`great_expectations.yml`): https://discuss.greatexpectations.io/t/environment-variable-substitution-is-not-working-for-me-when-connecting-ge-to-my-database/72

If there is something I missed please let me know!

Again, thank you for submitting this issue and for using Great Expectations!

For reference, here is the test I ran in a Windows environment to verify this is not due to an OS issue:
```
import great_expectations as ge
from great_expectations.data_context.types.base import (
    DataContextConfig,
    DataContextConfigSchema,
    DatasourceConfig,
    DatasourceConfigSchema,
)

kwargs = {
    "drivername": "snowflake",
    "host": "hostname",
    "warehouse": "compute_wh",
    "schema": "public",
    "database": "employee",
    "username": "username",
    "password": "${SF_PWD}"
}
data_ctxt = ge.DataContext.create('./', False)
data_ctxt.save_config_variable('SF', kwargs)

config_vars_file_contents = data_ctxt._load_config_variables_file()

# Make sure $ is escaped:
assert config_vars_file_contents["SF"]["password"] == r"\${SF_PWD}"

# Pull this into data context via added datasource
datasource_config = DatasourceConfig(
    class_name="SqlAlchemyDatasource",
    credentials="${SF}"
)
datasource_config_schema = DatasourceConfigSchema()
data_ctxt.add_datasource(
    initialize=False,
    name="test_datasource",
    **datasource_config_schema.dump(datasource_config)
)

# Assert datasources are listed as expected WITHOUT escaping
# (note password masking not yet merged in ge 0.13.3)
assert data_ctxt.list_datasources()[0]["credentials"] == kwargs

# Get data context config and make sure datasource substitution is as expected without `\`
data_context_config_schema = DataContextConfigSchema()
context_with_variables_substituted_dict = data_context_config_schema.dump(
    data_ctxt.get_config_with_variables_substituted()
)
test_datasource_credentials = context_with_variables_substituted_dict[
    "datasources"
]["test_datasource"]["credentials"]

assert test_datasource_credentials["password"] == "${SF_PWD}"
```

**Comment by varundunga** [comment]
Hi @anthonyburdi ! Thank you for looking at this issue. The config variables.yml that gets generated says that there are 2 types of 
substitutions possible and one of them is environment variable substitution. Moreover, this worked in previous versions of GE.
I think while adding the code to escape $ , this was not considered.
Also, the documentation in the below link says the same

[https://docs.greatexpectations.io/en/latest/guides/how_to_guides/configuring_data_contexts/how_to_use_a_yaml_file_or_environment_variables_to_populate_credentials.html](url)

**Comment by anthonyburdi** [comment]
@varundunga I just checked with the team and you are 100% correct. I'll make sure both the env var substitution in config_vars.yml and escaping of `$` characters works. I believe it may be due to the order in which it is carried out. Stay tuned and thank you for following up!

**Comment by varundunga** [comment]
@anthonyburdi thanks anthony. I have raised one PR for this issue(2220). May be you can take a look at that as well.

[https://github.com/great-expectations/great_expectations/pull/2220](url)

**Comment by anthonyburdi** [comment]
Hi @varundunga! Thanks again for bringing this issue up, reminding me of our feature set and proposing a solution in #2220. Since we support substituting variables using both the `${SOME_VAR}` and `$SOME_VAR` patterns, I had to check for both of those when saving config variables. I've incorporated that into #2243, added additional tests and have submitted for internal review (please also feel free to take a look if you like, comments are always appreciated :smile:). I'll close #2220 but am crediting the fix to you in our changelog via #2243. Thank you again.
